### PR TITLE
Allow UI layout updates without a window

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Allow UI layout updates without a window present.

## Solution

Make `LayoutContext` into a component that is used to mark an entity as a UI root entity with an associated Taffy root node and
add a `LayoutContext` component to the primary window entity in `UiPlugin::build`.

`ui_layout_system` now picks the first entity containing a `LayoutContext` component and introduces a corresponding taffy node to the Taffy layout tree (given that it isn't present already). This node is then set as the root of the entire tree. Then the UI layout update proceeds as normal.

---

## Changelog

changes:
* `LayoutContext` derives `Component`.  It has a new field `layout_to_logical_factor`, `scale_factor` is renamed to `combined_scale_factor`, `physical_size` is renamed to `root_node_size` and the `min_size` and `max_size` are removed.
* `UiPlugin::build` adds a `LayoutContext` component to the primary window entity.
* For the first `LayoutContext` entity, an associated Taffy node is added to the Taffy layout tree, which is then set as the root of the entire tree.
* Added a simple test that adds and then removes a UI node.

## Migration Guide
`LayoutContext` is now a component and its fields have been changed:
* `min` and `max` have been removed
* `scale_factor`  has been renamed to `combined_scale_factor`
* `layout_to_logical_factor` is a new field. The coordinates used in the internal UI layout tree are multiplied by `layout_to_logical_factor` after each UI update before being written to the `Node` and `Transform` components.

### CODE REVIEW

File: crates/bevy_ui/src/layout/mod.rs [hunk 1]
```diff
@@ -5,36 +5,34 @@ use crate::{ContentSize, Node, Style, UiScale};
...
 use bevy_math::Vec2;
 use bevy_transform::components::Transform;
 use bevy_utils::HashMap;
-use bevy_window::{PrimaryWindow, Window, WindowResolution, WindowScaleFactorChanged};
+use bevy_window::{PrimaryWindow, Window};
 use std::fmt;
 use taffy::{prelude::Size, style_helpers::TaffyMaxContent, Taffy};
 
+#[derive(Component, Debug, Copy, Clone)]
 pub struct LayoutContext {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/layout/mod.rs:24]: Basic doc strings while we're here please.

File: crates/bevy_ui/src/lib.rs [hunk 2]
```diff
@@ -119,7 +121,10 @@ impl Plugin for UiPlugin {
             .register_type::<widget::Label>()
             .add_systems(
                 PreUpdate,
-                ui_focus_system.in_set(UiSystem::Focus).after(InputSystem),
+                (
+                    ui_focus_system.in_set(UiSystem::Focus).after(InputSystem),
+                    ui_windows_system,
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/lib.rs:126]:
```suggestion
                    ui_windows_system.in_set(UiSystem::Window),
```

I'm not sure if there are other missing ordering constraints, but this is definitely implied by the doc strings above.

Comment (ickshonpe) [reply crates/bevy_ui/src/lib.rs]: Ah yeah, I wasn't sure whether to add a label to `UiSet` or not. `ui_window_system` runs in `PreUpdate` so there shouldn't be any need for ordering constraints.

**Comment by djeedai** [comment]
> Make ui_layout_system easier to test by moving the window logic into a separate system.

I disagree with that approach. Testing calls for a modular block of code, which means a function. Adding a separate system is a more heavyweight process that adds to the pile of things the scheduler has to do, and should only be done for runtime reasons. Can we move the logic into a function called from the same system it's currently in? That should solve the testability issue without requiring an additional system.

**Comment by ickshonpe** [comment]
> > Make ui_layout_system easier to test by moving the window logic into a separate system.
> 
> I disagree with that approach. Testing calls for a modular block of code, which means a function. Adding a separate system is a more heavyweight process that adds to the pile of things the scheduler has to do, and should only be done for runtime reasons. Can we move the logic into a function called from the same system it's currently in? That should solve the testability issue without requiring an additional system.

Ah, that's a good point. There is no need for a new system just for this change. The situation is a bit muddled and confused because it's not clear yet how multi ui root nodes and multi-window ui support will be implemented. I'll refactor this a bit.

**Comment by djeedai** [comment]
To be fair I'm on a phone so didn't look in details at the PR sorry. Just wanted to raise a general point. If that's too complicated for this case I'm fine using a separate system. It was just a suggestion in general, from experience I at least tend to put everything in a system function, and I think Bevy does that too, and that as you noted makes testing a lot harder. Separate non-system functions are nicer for testing but a pain to pass arguments to unfortunately.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by ickshonpe** [comment]
It was really helpful, I was making things over complicated. The implementation is much better now and realised I could easily hack together multiple window UI on top of this as well:

https://github.com/bevyengine/bevy/assets/27962798/52c120f1-7de6-4bcb-ae1d-9308b2354498

File: crates/bevy_ui/src/layout/mod.rs [hunk 3]
```diff
@@ -5,44 +5,69 @@ use crate::{ContentSize, Node, Style, UiScale};
...
+#[derive(Component, Debug, Reflect)]
+#[reflect(Component, Default)]
 pub struct LayoutContext {
-    pub scale_factor: f64,
-    pub physical_size: Vec2,
-    pub min_size: f32,
-    pub max_size: f32,
+    /// The size of the root node in the layout tree.
+    ///
+    /// Should match the size of the output window in physical pixels of the display device.
```
Comment (djeedai) [inline crates/bevy_ui/src/layout/mod.rs:31]: Is it a hard requirement that the root node covers the entire window? Or is it just the most common use case?

File: crates/bevy_ui/src/layout/mod.rs [hunk 4]
```diff
@@ -5,44 +5,69 @@ use crate::{ContentSize, Node, Style, UiScale};
...
+    /// After a UI layout has been computed, the layout coordinates are multiplied by `layout_to_logical_factor` to determine the final size of each UI Node entity to be stored in its [`Node`] component.
+    ///
+    /// `layout_to_logical_factor` is the reciprocal of the target window's `scale_factor` and doesn't include [`crate::UiScale`].
+    pub layout_to_logical_factor: f64,
 }
 
-impl LayoutContext {
-    /// create new a [`LayoutContext`] from the window's physical size and scale factor
-    fn new(scale_factor: f64, physical_size: Vec2) -> Self {
+impl Default for LayoutContext {
```
Comment (djeedai) [inline crates/bevy_ui/src/layout/mod.rs:43]: I'm a bit skeptical about this change. The `new()` was a lot more clear to me. Creating an implicit 800x600 layout feels very arbitrary (and undocumented).

File: crates/bevy_ui/src/layout/mod.rs [hunk 5]
```diff
@@ -57,7 +82,7 @@ impl fmt::Debug for UiSurface {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.debug_struct("UiSurface")
             .field("entity_to_taffy", &self.entity_to_taffy)
-            .field("window_nodes", &self.window_nodes)
+            .field("window_nodes", &self.root_nodes)
```
Comment (djeedai) [inline crates/bevy_ui/src/layout/mod.rs:85]: Isn't it better to rename the field too, since this is only for debug print anyway so we care little about back compatibility?

```suggestion
            .field("root_nodes", &self.root_nodes)
```

File: crates/bevy_ui/src/layout/mod.rs [hunk 6]
```diff
@@ -5,44 +5,69 @@ use crate::{ContentSize, Node, Style, UiScale};
...
-            max_size: physical_size.x.max(physical_size.y),
+            root_node_size: Vec2::new(800., 600.),
+            combined_scale_factor: 1.0,
+            layout_to_logical_factor: 1.0,
         }
     }
 }
 
+impl LayoutContext {
+    fn relative_ne(&self, other: &Self) -> bool {
```
Comment (djeedai) [inline crates/bevy_ui/src/layout/mod.rs:54]: We generally use `FloatOrd` to compare floats for equality in other places. I don't feel like it's justified to pull a new dependency just for having a slightly shorter syntax in that one function, while adding a new unfamiliar pattern for readers.

File: crates/bevy_ui/src/layout/mod.rs (hunk 3 above)
Comment (djeedai) [reply crates/bevy_ui/src/layout/mod.rs:31]: Reading below seems to suggest yes. Then I'd rephrase to be more assertive than "should".

```suggestion
    /// Set automatically during layout computation to the size of the output window in physical pixels of the display device.
```

**Review by djeedai — CHANGES_REQUESTED** [review]
Couple of things I'm not sure about, but fine to get overridden by any UI expert if they feel we can merge as is.

File: crates/bevy_ui/src/layout/mod.rs (hunk 6 above)
Comment (djeedai) [reply crates/bevy_ui/src/layout/mod.rs:54]: The new dependency also introduce a duplicate crate version and fails CI.

Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:54]: Yeah it's not needed anyway, I was forgetting how floats work, just partial equality is fine here/

File: crates/bevy_ui/src/layout/mod.rs [hunk 7]
```diff
@@ -5,44 +5,69 @@ use crate::{ContentSize, Node, Style, UiScale};
...
 use bevy_math::Vec2;
+use bevy_reflect::{std_traits::ReflectDefault, Reflect};
 use bevy_transform::components::Transform;
 use bevy_utils::HashMap;
-use bevy_window::{PrimaryWindow, Window, WindowResolution, WindowScaleFactorChanged};
+use bevy_window::Window;
 use std::fmt;
 use taffy::{prelude::Size, style_helpers::TaffyMaxContent, Taffy};
 
+/// Marks an entity as `UI root entity` with an associated root Taffy node and holds the resolution and scale factor information necessary to compute a UI layout.
```
[the remark this answers is not in the document]
Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:25]: It's a bit confusing because there are parts of the UI designed to support multiple windows but that hasn't been implemented completely and it's unclear what the API will be so I found writing these comments a bit difficult because I'm not sure how the final multi-window implementation will work. 

Maybe it just needs a note saying that the API isn't complete, may not make complete sense in places, and is likely to change drastically etc.

File: crates/bevy_ui/src/layout/mod.rs (hunk 3 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:31]: Yeah at the moment `ui_system_layout` sets the values for `LayoutContext` automatically when it's a component of a `Window` entity.

File: crates/bevy_ui/src/layout/mod.rs (hunk 4 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:43]: The default values will be overwritten by `ui_layout_system` with the window's values if running in a window. 

The default values are just there for convenience for the tests. You are right though, it might be confusing and seen as suggestive of something else. I'll just derive default.

**Comment by alice-i-cecile** [comment]
Can you add a simple migration guide for the renames?

**Comment by ickshonpe** [comment]
> Can you add a simple migration guide for the renames?

done

I didn't make one before because `LayoutContext` was added after the Bevy 10 release, do we need them for changes to features new to the next version?

**Comment by alice-i-cecile** [comment]
> I didn't make one before because LayoutContext was added after the Bevy 10 release, do we need them for changes to features new to the next version?

Ah I'd missed that. These are nice to include in the PR description still in case folks are working off main or the PR gets delayed. Ideally we'll exclude it from the migration guide notes but a false positive like that does no harm.

**Comment by ickshonpe** [comment]
Redundant, `UiTargetCamera` solved the same problem.
